curator-framework/src/main/java/org/apache/curator/framework/imps/EnsembleTracker.java

```diff
@@ -166,12 +166,26 @@ public static String configToConnectionString(QuorumVerifier data) throws Except
         StringBuilder sb = new StringBuilder();
         for ( QuorumPeer.QuorumServer server : data.getAllMembers().values() )
         {
+            if ( server.clientAddr == null )
+            {
+                // Invalid client address configuration in zoo.cfg
+                continue;
+            }
             if ( sb.length() != 0 )
             {
                 sb.append(",");
             }
-            InetSocketAddress address = Objects.firstNonNull(server.clientAddr, server.addr);
-            sb.append(address.getAddress().getHostAddress()).append(":").append(address.getPort());
+            InetAddress wildcardAddress = new InetSocketAddress(0).getAddress();
+            String hostAddress;
+            if ( wildcardAddress.equals(server.clientAddr.getAddress()) )
+            {
+                hostAddress = server.addr.getAddress().getHostAddress();
+            }
+            else
+            {
+                hostAddress = server.clientAddr.getAddress().getHostAddress();
+            }
+            sb.append(hostAddress).append(":").append(server.clientAddr.getPort());
```

Comment:
I don't believe getHostAddress() involves any DNS lookups, hostname resolution, etc. It just gets the String representation of the byte array on the InetAddressHolder. I may be mistaken about this though.